File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        return dfs(nums, 0, -1);
    }

    int dfs(vector<int>& nums, int i, int curr) {
        if (i >= nums.size()) {
            return 0;
        }

        if (curr == -1 || nums[i] > nums[curr]) {
            return max(dfs(nums, i + 1, curr), 1 + dfs(nums, i + 1, i));
        }
        else {
            return dfs(nums, i + 1, curr);
        }
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        def dfs(i, last, selling):
            if i >= len(prices):
                return 0
            
            if selling:
                return max(dfs(i + 1, last, True), prices[i] - prices[last] + dfs(i + 2, i, False))
            else:
                return max(dfs(i + 1, i, True), dfs(i + 1, last, False))
            
        return max(0, dfs(0, 0, False))
